### ISSUE TITLE
How do you feel about Shape Variables?

### ISSUE THREAD

**Issue by bionicles** [tensorflow/tensorflow#30925]
**System information**
- TensorFlow version (you are using): 2b1
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Ragged/variable shapes are currently described with (None, None, 10) type syntax. This is not informative for the user or the compiler.

How do you feel about the concept of Shape Variables? Shape syntax with strings or variables like ("batch_size", "n_atoms", 10) could be matched across the graph, potentially this could help with input signatures in tf.function and compiler code. 

this would be nice for reinforcement learning projects, and shape variables may be extremely useful in multi-task models and architecture search projects. Also this could improve explainability and interpretability of code

**Will this change the current api? How?**
Yes, users can specify variable "ragged" dimensions with strings or variables instead of **None**
**Who will benefit with this feature?**
Everyone who works on biochem, NLP, RL, multitask learning, and neural architecture search
**Any Other info.**
Just curious because we built a recursive NAS system for tf.keras in 2.0b1 this weekend and want to apply it to a bunch of different tasks, but they all have different shapes ... sometimes the shapes are variable and we need output shapes to match input shapes. "shape hell" sucks

**Comment by bionicles** [comment]
Closest thing I can think of to this is Strongly Typed Genetic Programming

![image](https://user-images.githubusercontent.com/24532336/61715882-07a60d80-ad2c-11e9-9508-993e99a622a0.png)

**Comment by bionicles** [comment]
@edloper 

just thinking about interfaces, we could use None, but then we don't have a good value for **Size**, which is sometimes needed

potentially what would work would be
to basically make 1 interface object per agent
and the interface creates "shape variables"

maybe it's not necessary but if we want to map from code_shape -> (None, 3), how do we know what to reshape to?
i guess we just do (-1, 3) ?
but then we need to make sure we have flattened size of 3N and we don't know N
but if it's (code_shape) -> (n_atoms, 3) then we have a way to calculate the size we need to decode the output

does this make sense?

**Comment by alextp** [comment]
It's a cool idea but a very large project. Many people have tried (see the contrib labeled_tensor for an attempt https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/labeled_tensor ) and none of this attempts was really functional.

Specifically, it's hard to preserve semantic names through operations which mix dimensions like convolutions and padding.

**Comment by bionicles** [comment]
@alextp @ymodak Yeah, it is a big project, but my concern is, None causes serious issues in the code, reads weirdly, and adds a bunch of boilerplate, like we have to write out all these Nones or it's gonna bug out. None means "nothing" -- but in Keras, it means "Anything" like a wildcard. TF/Keras "None" means the **_exact opposite_** of English "None"

If we need recurrent or recursive nets to produce a certain shape, there's no great way for the modules to figure out the shape for "None". The only way to really handle to "push" the shape onto the module (map over a placeholder) or track the shape variables in a lookup table, which is doable, but just adds more code.. 

Even the Input layer, automatically mutates shape to add None at the start. This shifts all the indices, and doesn't tell users, readers, or compiler anything. It's adding something, but the something is nothing, and nothing means anything? 

It doesn't need to be fancy. Plain python "int" type would be a lot more useful for 
everything with sequential data or variable shapes

```
batch_size = 4
n_atoms = None
input = K.Input((batch_size, n_atoms, 10))

later on...
output = tfpl.IndependentNormal((batch_size, n_atoms, 3))(x)  # just works
```
